akka-stream-tests/src/test/scala/akka/stream/impl/fusing/InterpreterSupervisionSpec.scala

```diff
@@ -0,0 +1,523 @@
+    var buf: List[Int] = Nil
+    def onPush(elem: Int, ctx: Context[Int]): Directive = {
+      buf = List(elem + 1, elem + 2, elem + 3)
+      ctx.push(elem)
+    }
+
+    override def onPull(ctx: Context[Int]): Directive = {
+      if (buf.isEmpty && ctx.isFinishing)
+        ctx.finish()
+      else if (buf.isEmpty)
+        ctx.pull()
+      else {
+        val elem = buf.head
+        buf = buf.tail
+        if (elem == 3) throw TE
+        ctx.push(elem)
+      }
+    }
+
+    override def onUpstreamFinish(ctx: Context[Int]): TerminationDirective =
+      if (absorbTermination)
+        ctx.absorbTermination()
+      else
+        ctx.finish()
+
+    // note that resume will be turned into failure in the Interpreter if exception is thrown from onPull
+    override def decide(t: Throwable): Supervision.Directive = decider(t)
+
+    override def restart(): OneToManyTestStage = copy()
+  }
+
+}
+
+class InterpreterSupervisionSpec extends InterpreterSpecKit {
+  import InterpreterSupervisionSpec._
+  import Supervision.stoppingDecider
+  import Supervision.resumingDecider
+  import Supervision.restartingDecider
+
+  "Interpreter error handling" must {
+
+    "handle external failure" in new TestSetup(Seq(Map((x: Int) ⇒ x + 1, stoppingDecider))) {
+      lastEvents() should be(Set.empty)
+
+      upstream.onError(TE)
+      lastEvents() should be(Set(OnError(TE)))
+
+    }
+
+    "emit failure when op throws" in new TestSetup(Seq(Map((x: Int) ⇒ if (x == 0) throw TE else x, stoppingDecider))) {
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
+      upstream.onNext(2)
+      lastEvents() should be(Set(OnNext(2)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
+      upstream.onNext(0) // boom
+      lastEvents() should be(Set(Cancel, OnError(TE)))
+    }
+
+    "emit failure when op throws in middle of the chain" in new TestSetup(Seq(
+      Map((x: Int) ⇒ x + 1, stoppingDecider),
+      Map((x: Int) ⇒ if (x == 0) throw TE else x + 10, stoppingDecider),
+      Map((x: Int) ⇒ x + 100, stoppingDecider))) {
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
+      upstream.onNext(2)
+      lastEvents() should be(Set(OnNext(113)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
+      upstream.onNext(-1) // boom
+      lastEvents() should be(Set(Cancel, OnError(TE)))
+    }
+
+    "resume when Map throws" in new TestSetup(Seq(Map((x: Int) ⇒ if (x == 0) throw TE else x, resumingDecider))) {
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
+      upstream.onNext(2)
+      lastEvents() should be(Set(OnNext(2)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
+      upstream.onNext(0) // boom
+      lastEvents() should be(Set(RequestOne))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
```

Comment:
@drewhk Should two not be impossible? After the “boom” the ball landed upstream so we should just feed more data, right?